Adding METADATA support

Sergey-Kitov opened the pull request:
Duplicate of PR to jonhoo/rust-imap

Commit:
Adding METADATA support

hpk42 commented:
could you re-do this PR to async-email/async-imap repo? 

On Tue, Nov 12, 2019 at 05:07 -0800, Sergey-Kitov wrote:
> Duplicate of PR to jonhoo/rust-imap
> You can view, comment on, or merge this pull request online at:
> 
>   https://github.com/deltachat/rust-imap/pull/3
> 
> -- Commit Summary --
> 
>   * Adding METADATA support
> 
> -- File Changes --
> 
>     M Cargo.toml (2)
>     M src/client.rs (14)
>     A src/extensions/metadata.rs (166)
>     M src/extensions/mod.rs (1)
>     M src/lib.rs (2)
>     M src/parse.rs (27)
>     M src/types/capabilities.rs (34)
>     M src/types/mod.rs (9)
> 
> -- Patch Links --
> 
> https://github.com/deltachat/rust-imap/pull/3.patch
> https://github.com/deltachat/rust-imap/pull/3.diff
> 
> -- 
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/deltachat/rust-imap/pull/3

Sergey-Kitov commented:
Async-std is not compiling, how do you use async-imap?

hpk42 commented:
We use async-imap in https://github.com/deltachat/deltachat-core-rust/
and it compiles fine with nightly-2019-11-06.

Sergey-Kitov commented:
Did you mean deltachat/async-imap? for me it is not compiling due to compilation error in async-std, not in async-imap itself

hpk42 commented:
On Wed, Nov 27, 2019 at 00:57 -0800, Sergey-Kitov wrote:
> Did you mean deltachat/async-imap? for me it is not compiling due to compilation error in async-std

ah, no.  that was an old fork -- i just removed that repo to avoid the confusion. 

compiling should work with this: 

https://github.com/async-email/async-imap/

if not, please open an issue there.

Sergey-Kitov commented:
So the problem was in rust version...

Celti committed:
Add `imap::connect_starttls` convenience function

Add a `imap::connect_starttls` convenience function, with the same
syntax as `imap::connect`, to make STARTTLS connections as easily as
IMAPS.

PR#140 removed `imap::connect_insecure` with the stated goal of
encouraging users to connect securely. With that change, users are
forced to construct their own `TcpStream`s when interacting with
non-IMAPS servers.

This change may perversely incentivise the use of insecure connections,
instead of discouraging them, as users may blindly copy-paste code
involving `Client::new` and `TcpStream`s so things "just work", without
the visual indicator of `_insecure` suggesting something is wrong.

Tangentially, [RFC 2595] encourages using STARTTLS instead of raw SSL
sockets. Ideally, we should support both options equally well.

[RFC 2595]: https://tools.ietf.org/html/rfc2595#section-7

jonhoo committed:
Merge pull request #153 from Celti/starttls

Add `imap::connect_starttls` convenience function

Commit:
Ignore server comments

jonhoo committed:
Fix up CI to use new fancy Azure Pipelines CI

jonhoo committed:
Merge pull request #155 from jonhoo/fix-ci

Fix CI

jonhoo committed:
Merge pull request #154 from dmeijboom/master

When authenticating using SASL XAUTH2 with Gmail fails

jonhoo committed:
Bump dependencies

This also pulls in imap-proto 0.11, which exposes message envelopes as
`&[u8]` instead of `&str`. This directly affects our public API.

jonhoo committed:
Bump version for breaking release

This includes the imap-proto 0.11 `&str` -> `&[u8]` change, and the
`Capability` changes in imap-proto 0.10 (and #133).

jonhoo committed:
Merge pull request #156 from jonhoo/breaking-release

Prepare breaking release (2.0)

mordak committed:
Make Display implementation independent of deprecated Error::description.

Allow use of deprecated in existing Error::description implementation to
suppress warnings.

jonhoo committed:
Allow errors on nightly

jonhoo committed:
Merge pull request #158 from mordak/error-display

Make Display implementation independent of Error::description

jonhoo committed:
Bump dependencies

jonhoo committed:
Bump version for new Error impls (#158)

mredaelli committed:
Support timing out wait on IDLE

mordak committed:
Handle Dovecot's periodic IDLE notification messages ("OK Still here").

These messages show up at an interval controlled by Dovecot's
imap_idle_notify_interval setting, which defaults to 2 minutes.
These messages were causing the IDLE loop to exit prematurely
because it looks like a notification response from the server.

jonhoo committed:
Merge pull request #160 from mordak/idle-ok-still-here

Handle Dovecot's periodic IDLE notification messages ("OK Still here").

jonhoo committed:
Bump version for Dovecot IDLE fix

jonhoo committed:
Merge pull request #159 from mredaelli/master

Support timing out wait on IDLE

jonhoo committed:
Fix BSD ci

jonhoo committed:
Bump outdated dev-dep

jonhoo committed:
Bump version for new, richer wait_with_timeout

jonhoo committed:
FLAGS can be sent as unilateral response

Fixes #161 (well, the side-issue in #161).

jonhoo committed:
Merge pull request #162 from jonhoo/unilateral-flags

FLAGS can be sent as unilateral response

jonhoo committed:
Release with handling of unilateral FLAGS

jonhoo committed:
Expunge holds a Seq

jonhoo committed:
Use type alias for better docs

jonhoo committed:
Expose unilateral mailbox flag changes

This is a backwards incompatible change, since it adds a variant to a
public enum.

jonhoo committed:
Make UnsolicitedResponse non_exhaustive

That way, as we discover additional unilateral responses in the future,
we won't have to make a breaking change to add them.

jonhoo committed:
Bump MSRV for non_exhaustive

rvuongdav committed:
Changed: a typo was fixed

jonhoo committed:
Merge pull request #164 from rvuong/fixes-a-minor-typo

Changed: a typo was fixed

rvuongdav committed:
Added: STARTTLS example

rvuongdav committed:
Changed: description update, plus code formatting

jonhoo committed:
Merge pull request #165 from rvuong/added-starttls-example

Added: STARTTLS example

wookietreiber committed:
Add note about server greeting to Client::new

wookietreiber committed:
Add timeout example

jonhoo committed:
Bump MSRV

jonhoo committed:
Update remote CI deps

jonhoo committed:
Merge pull request #169 from jonhoo/ci-fixes

Make CI pass on `master` again

jonhoo committed:
Merge pull request #168 from wookietreiber/greeting-and-timeout

server greeting doc and timeout example

timando committed:
Re-export Result and Error types (#170)

jonhoo committed:
Release 2.2.0

mordak committed:
Add append_with_flags (#171)

Allows a client to set flags on a new message as it is appended to a mailbox.

jonhoo committed:
Release 2.3.0 with append_with_flags

mordak committed:
Fix trivial clippy warnings.

mordak committed:
Fix clippy manual_non_exhaustive lint.

mordak committed:
Update imap-proto and nom dependencies.

Add support for HIGHESTMODSEQ (RFC 4551) and VANISHED (RFC 7162),
which allows users to quickly synchronize to a mailbox by fetching
only changes since the last known highest mod sequence.

mordak committed:
Fix typo in doc comment

Co-authored-by: Jon Gjengset <jon@thesquareplanet.com>

mordak committed:
Add non_exhaustive to Mailbox.

mordak committed:
Add a Deleted type to wrap EXPUNGE and VANISHED responses.

EXPUNGE may return either a series of EXPUNGE responses each with
a single message sequence number, or a VANISHED response with a
sequence set of UIDs. This adds a wrapper enum and some associated
iterators to make it easy to handle these in the client.

mordak committed:
Bump minimum version.

Via nom6 dependency bitvec.

mordak committed:
Update CHANGELOG

mordak committed:
Link to QRESYNC RFC section describing VANISHED vs EXPUNGE responses.

mordak committed:
CI: Only install components when needed. (#173)

mblarsen committed:
feat: allow setting sent date on APPEND (#174)

Fixes #60

jonhoo committed:
Bump version for append_with_date

mblarsen committed:
refactor: combine all append_* + introduce AppendOptions

mblarsen committed:
feat: add AppendCmd builder

mblarsen committed:
refactor: use AppendCmd instead of AppendOptions

mblarsen committed:
fix: pass session as &mut

mblarsen committed:
fix: correct lifetimes and types for append

mblarsen committed:
feat: add 'flags' method to add multiple flags at once

mblarsen committed:
docs: move flag + date documentation to AppendCmd

mblarsen committed:
refactor: use extend instead of append

Co-authored-by: Jon Gjengset <jon@thesquareplanet.com>

mblarsen committed:
refactor: move must_use to AppendCmd

mblarsen committed:
refactor: rename run to finish

mblarsen committed:
docs: improve docs

mblarsen committed:
test: correct test for append_with_flags

mmirate committed:
Avoid trying to FETCH an empty set of messages (#177)

Also, apply correct validation to FETCH arguments.

jonhoo committed:
Release 2.4.1

jonhoo committed:
Prepare for 3.0.0 alpha

jonhoo committed:
Merge branch 'master' into more-unilaterals

jonhoo committed:
Merge pull request #163 from jonhoo/more-unilaterals

More unilaterals

jonhoo committed:
Merge pull request #175 from mblarsen/refactor/squash-append-cmd

refactor: squash append command

jonhoo committed:
Merge branch 'master' into mod_seq_vanished

jonhoo committed:
Merge pull request #172 from mordak/mod_seq_vanished

Support HIGHESTMODSEQ and VANISHED

jonhoo committed:
Merge branch 'master' into metadata-merge
